### PR TITLE
Add per_tensor_quantize to int8 quantize

### PR DESCRIPTION
**Pull request by zihaomu**
Hi, the purpose of this PR is to add per-tensor quantization to the model quantization part of opencv dnn. 
[The difference of per-channel quantization and per-tensor quantization](https://www.tensorflow.org/lite/performance/quantization_spec).

The existing quantization method in opencv/dnn is based on **per-channel quantization**, which can achieve better accuracy than **per-tensor quantization**. But for some hardware, per-tensor quantization will be easier to optimize the speed, especially NPU chips.

For example, the [NPU of TIM-VX](https://github.com/VeriSilicon/TIM-VX) backend:
ResNet50 int8 (per-channel) takes **525.298 ms** on Khadas Vim3,
And ResNet50 int8 (per-tensor) takes **20.01 ms** on Khadas Vim3.

Compared with per-channel quantize, the per-tensor quantization has the disadvantage of low accuracy. Some original unit test in `dnn/test/test_int8_layer.cpp` cannot be passed due to low accuracy. So I changed the threshold of the parameter in some unit test cases.

[Related PR](https://github.com/opencv/opencv/pull/21036).

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] The feature is well documented and sample code can be built with the project CMake
- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.

### CODE REVIEW

**Comment by vpisarev** [comment]
:+1:
@zihaomu, please, resolve merge conflicts and this PR can be merged

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 1]
```diff
@@ -601,14 +601,22 @@ CV__DNN_INLINE_NS_BEGIN
          */
         CV_WRAP_AS(forwardAndRetrieve) void forward(CV_OUT std::vector<std::vector<Mat> >& outputBlobs,
                                                     const std::vector<String>& outBlobNames);
-
```
Comment (rogday) [inline modules/dnn/include/opencv2/dnn/dnn.hpp:604]: I think it was removed by accident.

File: modules/dnn/src/layers/fully_connected_layer.cpp [hunk 2]
```diff
@@ -621,21 +621,48 @@ class FullyConnectedLayerImpl CV_FINAL : public InnerProductLayer
         Mat outputMultiplier(1, numOutput, CV_32F);
 
         double realMin, realMax, weightsScale;
-        for( int i = 0; i < numOutput; i++ )
+
+        bool perTensor = false;
+        if(params.has("perTensor"))
+            perTensor = params.get<bool>("perTensor");
```
Comment (rogday) [inline modules/dnn/src/layers/fully_connected_layer.cpp:627]:
```cpp
bool perTensor = params.get<bool>("perTensor", false);
```

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 3]
```diff
@@ -2173,22 +2173,47 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
...
+                weightsMat.row(i).convertTo(weightsQuantized.row(i), CV_8S, 1.f/weightsScale);
+
+                // Quantize biases
+                float biasScale = inputScale * weightsScale;
+                biasQuantized.at<int>(i) = (int)std::round(biasvec[i]/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
+
+                // Store multiplier
+                outputMultiplier.at<float>(i) = biasScale / outputScale;
+            }
         }
```
[the remark this answers is not in the document]
Comment (zihaomu) [reply modules/dnn/src/layers/convolution_layer.cpp:2217]: Thanks for code reviewing. I will try to improve the code based on your comment.

**Comment by zihaomu** [comment]
@rogday I have removed some duplicating code, but not that much.

**Comment by zihaomu** [comment]
> I was thinking more of along [these](https://github.com/rogday/opencv/commit/4a00dec78f7d420d25a28c5ff3604686364a5926) lines. Take a look, feel free to disagree. It has major disadvantage of taking a lot of parameters. But no code duplication.

Thanks for your reply. My concern is a function with a lot of parameters maybe make code more complicated and hard to understand.

**Comment by alalek** [comment]
Please rebase to resolve merge conflict: `modules/dnn/test/test_int8_layers.cpp`

**Comment by asmorkalov** [comment]
@zihaomu @rogday Friendly reminder.

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 4]
```diff
@@ -2173,23 +2173,33 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
...
-            // Quantize biases
+            weightsMat.convertTo(weightsQuantized, CV_8S, 1.f/weightsScale);
             float biasScale = inputScale * weightsScale;
-            biasQuantized.at<int>(i) = (int)std::round(biasvec[i]/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
 
-            // Store multiplier
-            outputMultiplier.at<float>(i) = biasScale / outputScale;
+            for (int i = 0; i < numOutput; i++)
+            {
+                biasQuantized.at<int>(i) = (int)std::round(biasvec[i]/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
```
Comment (rogday) [inline modules/dnn/src/layers/convolution_layer.cpp:2188]: `static_cast<int>(...)`

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 5]
```diff
@@ -2173,23 +2173,33 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
...
+        }
+        else // per Channel
+        {
+            for (int i = 0; i < numOutput; i++)
+            {
+                double weightsScale = getWeightScale(weightsMat.row(i));
+
+                weightsMat.row(i).convertTo(weightsQuantized.row(i), CV_8S, 1.f/weightsScale);
+                float biasScale = inputScale * weightsScale;
+                biasQuantized.at<int>(i) = (int)std::round(biasvec[i]/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
```
Comment (rogday) [inline modules/dnn/src/layers/convolution_layer.cpp:2200]: `static_cast<int>(...)`

File: modules/dnn/src/layers/fully_connected_layer.cpp [hunk 6]
```diff
@@ -619,23 +619,33 @@ class FullyConnectedLayerImpl CV_FINAL : public InnerProductLayer
...
+
+            weightsMat.convertTo(weightsQuantized, CV_8S, 1.f/weightsScale);
             float biasScale = inputScale * weightsScale;
-            biasQuantized.at<int>(i) = (int)std::round(biasMat.at<float>(i)/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
 
-            // Store multiplier
-            outputMultiplier.at<float>(i) = biasScale / outputScale;
+            for (int i = 0; i < numOutput; i++)
+            {
+                biasQuantized.at<int>(i) = (int)std::round(biasMat.at<float>(i)/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
```
Comment (rogday) [inline modules/dnn/src/layers/fully_connected_layer.cpp:634]: `static_cast<int>(...)`

File: modules/dnn/src/layers/fully_connected_layer.cpp [hunk 7]
```diff
@@ -619,23 +619,33 @@ class FullyConnectedLayerImpl CV_FINAL : public InnerProductLayer
...
+        }
+        else
+        {
+            for (int i = 0; i < numOutput; i++)
+            {
+                double weightsScale = getWeightScale(weightsMat.row(i));
+
+                weightsMat.row(i).convertTo(weightsQuantized.row(i), CV_8S, 1.f/weightsScale);
+                float biasScale = inputScale * weightsScale;
+                biasQuantized.at<int>(i) = (int)std::round(biasMat.at<float>(i)/biasScale) - inputZp*(cv::sum(weightsQuantized.row(i))[0]);
```
Comment (rogday) [inline modules/dnn/src/layers/fully_connected_layer.cpp:646]: `static_cast<int>(...)`

**Review by rogday — CHANGES_REQUESTED** [review]
Looks good. We have a few duplicating lines, but given the alternative - function with lots of parameters, it's fine.

File: modules/dnn/src/layers/fully_connected_layer.cpp (hunk 7 above)
Comment (zihaomu) [reply modules/dnn/src/layers/fully_connected_layer.cpp:646]: Thanks for the code review, I will revise the code and submit it soon.

**Comment by asmorkalov** [comment]
@zihaomu Friendly reminder.

**Comment by asmorkalov** [comment]
@zihaomu Friendly reminder.

**Comment by zihaomu** [comment]
@asmorkalov Thanks for the reminder, at the moment I'm all into [fast Conv](https://github.com/opencv/opencv/pull/21910). I'll be updating this PR early next week.

**Comment by zihaomu** [comment]
Hi @rogday, the code has been updated.

**Review by rogday — APPROVED** [review]
LGTM! :+1:

File: modules/dnn/src/net_quantization.cpp [hunk 8]
```diff
@@ -32,8 +32,13 @@ void getQuantizationParams(const Mat& src, std::vector<float>& scales, std::vect
     zeropoints.push_back((int)std::round(zp));
 }
 
-// FIXIT drop from inference API
 Net Net::Impl::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype)
+{
+    return quantize(calibData, inputsDtype, outputsDtype, false);
+}
```
Comment (alalek) [inline modules/dnn/src/net_quantization.cpp:38]: Do we need this method? (in internal class)

File: modules/dnn/test/test_int8_layers.cpp [hunk 9]
```diff
@@ -1269,6 +1308,9 @@ TEST_P(Test_Int8_nets, YOLOv4_tiny)
     {
         SCOPED_TRACE("batch size 1");
         testDarknetModel(config_file, weights_file, ref.rowRange(0, N0), scoreDiff, iouDiff, confThreshold);
+
+        // per-tensor quantize
+        testDarknetModel(config_file, weights_file, ref.rowRange(0, N0), scoreDiff, 0.16, 0.7, 0.4, true);
```
Comment (alalek) [inline modules/dnn/test/test_int8_layers.cpp:1313]: BTW, It makes sense to create dedicated test or at least use `SCOPED_TRACE();`.
(here and above)

File: modules/dnn/src/net_quantization.cpp (hunk 8 above)
Comment (zihaomu) [reply modules/dnn/src/net_quantization.cpp:38]: Thanks for code reviewing. 
How about removing `Net Net::Impl::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype)` and just keeping `Net Net::Impl::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype, bool perTensor = false)`?

File: modules/dnn/test/test_int8_layers.cpp (hunk 9 above)
Comment (zihaomu) [reply modules/dnn/test/test_int8_layers.cpp:1313]: Ok, I will try to update it.

File: modules/dnn/src/net_quantization.cpp (hunk 8 above)
Comment (alalek) [reply modules/dnn/src/net_quantization.cpp:38]: We could remove this if it is not used anymore.

**Comment by zihaomu** [comment]
@alalek There is a API ERROR, due to we have removed the oringinal `Net Net::Impl::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype)`. [CI Report](https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/100477/steps/Compare%20ABI%20dumps/logs/report-html).

**Comment by alalek** [comment]
>  due to we have removed the oringinal `Net Net::Impl::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype)`

`Net::Impl` is internal API. It is not checked by API compatibility tool.

Tool blames on public `Net::quantize ( cv::InputArrayOfArrays calibData, int inputsDtype, int outputsDtype )` (we still need thin wrapper for it over extended overload)

---

I believe we could add this call into "skip" list of this tool as quantization API is experimental (and we don't need to maintain compatibility overloads here).
I will do this later if there is no objections.

**Comment by vpisarev** [comment]
@zihaomu, thank you! What about the case when we already have quantized model stored in ONNX format and load it - does it recognize "per-tensor" case?

**Comment by zihaomu** [comment]
> @zihaomu, thank you! What about the case when we already have quantized model stored in ONNX format and load it - does it recognize "per-tensor" case?

This PR only affects the fly-quantize model. And about per-quantized ONNX model, if the original model is `per-tensor` model, it will be run in `per-tensor` way. And if it is `per-channel` model, it will run in the `per-channel` way.

**Comment by zihaomu** [comment]
Hi @vpisarev @alalek and @rogday, I have changed the API from `perTensor` to `perChannel`. Because the ONNX uses the same Quantize Flag at the [quantize_static(..., per_channel, ...)](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/quantize.py). In addition, the `onnx_importer` can recognize quantization type at `QConv` and `QMatMul` Layers.

File: modules/dnn/src/net.cpp [hunk 10]
```diff
@@ -115,12 +115,12 @@ void Net::forward(std::vector<std::vector<Mat>>& outputBlobs,
 }
 
 // FIXIT drop from inference API
-Net Net::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype)
+Net Net::quantize(InputArrayOfArrays calibData, int inputsDtype, int outputsDtype, bool perTensor = false)
```
Comment (alalek) [inline modules/dnn/src/net.cpp:118]: > bool perTensor = false

Default value works properly from .hpp files only. They are useless in .cpp files

Comment (zihaomu) [reply modules/dnn/src/net.cpp]: Thanks, fixed.

**Review by vpisarev — APPROVED** [review]

**Pull request merged**